### ISSUE TITLE
ERROR: CREATE INDEX CONCURRENTLY cannot be executed within a pipeline [ERROR:0, SQLSTATE:25001]

### ISSUE THREAD

**Issue by ljdelight** [maproulette/maproulette-backend#1019]
During startup of a maproulette instance using postgres 13.10, I see this error: `ERROR: CREATE INDEX CONCURRENTLY cannot be executed within a pipeline [ERROR:0, SQLSTATE:25001]`. There is no error when using postgres 13.9.

This was first noticed with CI/CD GitHub actions and is easily repro'd on a local database This appears to be a play framework evolution that doesn't work well with the latest postgres 13 drop.

---
This is one way to start a clean database locally using the latest pg13:
```sh
podman run \
    -d \
    -p 5432:5432 \
    --pull=always \
    --name maproulette-postgis2 \
    --restart unless-stopped \
    --shm-size=2048MB \
    -e POSTGRES_DB=maproulette-db \
    -e POSTGRES_USER=maproulette-db-user \
    -e POSTGRES_PASSWORD=maproulette-db-pass \
    postgis/postgis:13-3.3
```

---

After starting the database and starting MapRoulette within intellij, the error log shows

`22:29:05,771 [error] [play-dev-mode-akka.actor.default-dispatcher-8] p.a.d.e.DatabaseEvolutions.$anonfun$checkEvolutionsState$3(EvolutionsApi.scala:302) - ERROR: CREATE INDEX CONCURRENTLY cannot be executed within a pipeline [ERROR:0, SQLSTATE:25001]`

and after opening the webUI (http://127.0.0.1:9000/docs) this is displayed:
```
We got the following error: ERROR: CREATE INDEX CONCURRENTLY cannot be executed within a pipeline [ERROR:0, SQLSTATE:25001], while trying to run this SQL script:

1  -- Rev:48,Ups - 313b22e
2
3  -- Setup constraints on virtual_project_challenges to restrict duplicate
4  -- entries.
5
6  -- Find and fix prior duplicate entries.
7  DELETE FROM
8  virtual_project_challenges a
9  USING virtual_project_challenges b
10 WHERE
11 a.id < b.id
12 AND a.project_id = b.project_id
13 AND a.challenge_id = b.challenge_id;;
14
15 -- Add unique constraint.
16 CREATE UNIQUE INDEX CONCURRENTLY virtual_project_challenges_projects
17 ON virtual_project_challenges (project_id, challenge_id);;
18
19 ALTER TABLE virtual_project_challenges DROP CONSTRAINT
20 IF EXISTS unique_virtual_project_challenges_projects;;
21 ALTER TABLE virtual_project_challenges
22 ADD CONSTRAINT unique_virtual_project_challenges_projects
23 UNIQUE USING INDEX virtual_project_challenges_projects;;

```

**Comment by ljdelight** [comment]
After some testing, I confirmed that this is NOT limited to a "clean/new" database. This error will happen when using any postgres version that doesn't like the 'CONCURRENTLY' in the above statement.

**Comment by ljdelight** [comment]
Another tool had the same error as maproulette and they reached out to the pg dev group:

> 
> https://www.postgresql.org/message-id/4139340.1676050723%40sss.pgh.pa.us
> Timur <timur(dot)luchkin(at)gmail(dot)com> writes:
> > After the release of PostgreSQL 14.7 Liquibase started to fail with:
> 
> 
> > Reason: liquibase.exception.DatabaseException: ERROR: CREATE INDEX 
> > CONCURRENTLY cannot be executed within a pipeline [Failed SQL:
> 
> 
> You need to report that to the Liquibase authors.  They're
> trying to do something that's unsupported, but which we previously
> failed to detect.  Basically: you can't do that in a pipeline
> because there will already be an open transaction, which would
> accidentally get committted by the CREATE INDEX CONCURRENTLY.
> 
> 			regards, tom lane

Removing 'CONCURRENTLY' does resolve the issue. I'll do some more testing to see impact to evolutions on an existing system.

**Comment by ljdelight** [comment]
MapRoulette Version: v4.3.1
PostgreSQL Version: 11.19, 12.14, 13.10, 14.7, 15.2

PostgreSQL had an update that improves validation that the previous versions failed to check, and this causes MapRoulette backend v4.3.1 and lower to fail to start with error `ERROR: CREATE INDEX CONCURRENTLY cannot be executed within a pipeline [ERROR:0, SQLSTATE:25001]`.

- NOTE: This error is not going to happen when using an older / outdated version. This error happens with PostgreSQL Version: 11.19, 12.14, 13.10, 14.7, 15.2.

## Resolution

The `CONCURRENTLY` syntax was removed from MapRoulette evolution `conf/evolutions/default/48.sql` and this resolves the startup issue. However, because of how Play evolutions work for production mode, **MapRoulette deployments will fail at startup and will require manual remediation.**

### What do I need to do?

When updating from MapRoulette version _less than or equal to_ v4.3.1 to a later version, use the below SQL updates to manually remediate the evolution differences.


```sql
DELETE FROM public.play_evolutions WHERE id = 48;

INSERT INTO public.play_evolutions VALUES (48, '876ad95c8b73eb7d08fdaf24b21399d5c01918d2', '2023-02-26 21:05:52.454', '-- Setup constraints on virtual_project_challenges to restrict duplicate
-- entries.

-- Find and fix prior duplicate entries.
DELETE FROM
virtual_project_challenges a
USING virtual_project_challenges b
WHERE
a.id < b.id
AND a.project_id = b.project_id
AND a.challenge_id = b.challenge_id;;

-- Add unique constraint.
CREATE UNIQUE INDEX virtual_project_challenges_projects
ON virtual_project_challenges (project_id, challenge_id);;

ALTER TABLE virtual_project_challenges DROP CONSTRAINT
IF EXISTS unique_virtual_project_challenges_projects;;
ALTER TABLE virtual_project_challenges
ADD CONSTRAINT unique_virtual_project_challenges_projects
UNIQUE USING INDEX virtual_project_challenges_projects;;', 'ALTER TABLE virtual_project_challenges
DROP CONSTRAINT IF EXISTS unique_virtual_project_challenges;;', 'applied', '');
```

### What do I do if I need to revert back?

If you need to rollback to MapRoulette version `<= v4.3.1` for whatever reason, use the below manual step to replace the `CONCURRENTLY` query.

```sql
DELETE FROM public.play_evolutions WHERE id = 48;

INSERT INTO public.play_evolutions VALUES (48, '313b22e7601c3875b12416f54b47c6c754e16af2', '2023-02-04 22:40:17.893', '-- Setup constraints on virtual_project_challenges to restrict duplicate
-- entries.

-- Find and fix prior duplicate entries.
DELETE FROM
virtual_project_challenges a
USING virtual_project_challenges b
WHERE
a.id < b.id
AND a.project_id = b.project_id
AND a.challenge_id = b.challenge_id;;

-- Add unique constraint.
CREATE UNIQUE INDEX CONCURRENTLY virtual_project_challenges_projects
ON virtual_project_challenges (project_id, challenge_id);;

ALTER TABLE virtual_project_challenges DROP CONSTRAINT
IF EXISTS unique_virtual_project_challenges_projects;;
ALTER TABLE virtual_project_challenges
ADD CONSTRAINT unique_virtual_project_challenges_projects
UNIQUE USING INDEX virtual_project_challenges_projects;;', 'ALTER TABLE virtual_project_challenges
DROP CONSTRAINT IF EXISTS unique_virtual_project_challenges;;', 'applied', '');
```

**Comment by ljdelight** [comment]
Merged the syntax fix in `v4.3.3` and the above instructions worked great. Closing.
